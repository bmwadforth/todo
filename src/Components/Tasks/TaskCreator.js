import React, {useState} from "react";
import uuid from "uuid/v4";
import PropTypes from "prop-types";
import Field from "../Form/Field";
import WebForm from "../Form/WebForm";
import {useDispatch, useSelector} from "react-redux";
import {ACTIONS} from "../../Actions/Actions";
import {TaskExists} from "../../Actions/TaskActions";
import Alert from "../Utility/Alert";

export default function TaskCreator(props) {
    const [taskCreated, setTaskCreated] = useState(false);
    const [error, setError] = useState(null);
    const [task, setTask] = useState({
        id: null,
        title: "",
        description: "",
        cutOff: new Date(),
        created: new Date(),
        completed: false
    });
    const dispatch = useDispatch();
    const {data: tasks} = useSelector(store => store.Tasks);

    function newTask(e) {
        e.preventDefault();
        if (TaskExists(task, tasks)) {
            setError("Task Already Exists");
            setTaskCreated(false);
        } else {
            dispatch({type: ACTIONS.TASK.CREATE, payload: {...task, id: uuid()}});
            setError(null);
            setTaskCreated(true);
        }
    }

    return (
        <div className="task-creator">
            {taskCreated && <Alert title="New Task Added" status="success"/>}
            {error && <Alert title={error} status="danger"/>}
            <WebForm actionTitle="New Task" onSubmit={newTask}>
                <Field id="task-title" title="Task Title" onChange={e => {
                    setTask({...task, title: e.target.value});
                }} value={task.title} required={true}/>

                <Field id="task-desc" title="Task Description" onChange={e => {
                    setTask({...task, description: e.target.value});
                }} value={task.description} required={false}/>

                <Field id="task-cutOff" title="Task Due Date" onChange={e => {
                    setTask({...task, cutOff: e.target.value});
                }} value={task.cutOff} type="date" required={true}/>
            </WebForm>
        </div>
    );
}

TaskCreator.propTypes = {};
