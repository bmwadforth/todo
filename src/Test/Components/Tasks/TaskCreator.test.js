import React from 'react';
import {mount, shallow, configure} from "enzyme";
import Adapter from 'enzyme-adapter-react-16';
import TaskCreator from "../../../Components/Tasks/TaskCreator";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
configure({adapter: new Adapter()});

describe("TaskCreator Component", () => {
    const initialState = {Tasks: {data: {}}};
    const mockStore = configureStore();
    let store;

    beforeEach(() => {
        store = mockStore(initialState);
    });

    it("Should Match Snapshot", () => {
        const wrapper = shallow(<Provider store={store}><TaskCreator /></Provider>);
        expect(wrapper.html()).toMatchSnapshot();
    });
});
