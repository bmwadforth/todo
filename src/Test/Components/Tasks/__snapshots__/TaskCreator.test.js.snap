// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskCreator Component Should Match Snapshot 1`] = `"<div class=\\"task-creator\\"><h2>Create Task</h2><form><div class=\\"web-form\\"><div class=\\"form-item\\"><input type=\\"text\\" id=\\"task-title\\" value=\\"\\" placeholder=\\"Task Title\\" required=\\"\\"/></div><div class=\\"form-item\\"><input type=\\"text\\" id=\\"task-desc\\" value=\\"\\" placeholder=\\"Task Description\\"/></div><button class=\\"button button-primary\\" type=\\"submit\\">Add Task</button></div></form></div>"`;
